charmhub_lp_tools/osci_sync.py

```diff
@@ -0,0 +1,247 @@
+from typing import (
+    Any,
+    Dict,
+    List,
+    Tuple,
+)
+
+import git
+import yaml
+
+from .charm_project import (
+    DEFAULT_RECIPE_FORMAT,
+)
+from .constants import (
+    LIST_AUTO_BUILD_CHANNELS,
+    OSCI_YAML,
+)
+from .exceptions import (
+    BranchNotFound,
+    CharmNameNotFound,
+)
+from .gitutils import (
+    get_branch_name,
+)
+from .group_config import GroupConfig
+from .parsers import (
+    parse_channel,
+)
+
+
+OsciYamlType = List[Any]
+logger = logging.getLogger(__name__)
+
+
+def diff_auto_build_channels(lp_auto_build_channels: Dict[str, str],
+                             local_auto_build_channels: Dict[str, str]):
+    """
+    Obtain the differences between auto build channels maps.
+
+    The differences are obtained using the symmetic difference operation from
+    sets.
+
+    :param lp_auto_build_channels: the baseline auto build channels as defined
+                                   in Launchpad build recipe
+    :param local_auto_build_channels: the auto build channels built from
+                                      information defined in osci.yaml
+    :returns: the differences between both auto build channels map.
+    """
+    a_set = set(lp_auto_build_channels.items())
+    b_set = set(local_auto_build_channels.items())
+    return dict(a_set ^ b_set)
+
+
+def find_osci_yaml(git_repo: git.Repo) -> pathlib.Path:
+    """Find the osci.yaml file path.
+
+    :param git_repo: git repo of the charm
+    :return: The path to the osci.yaml file.
+    :raise FileNotFoundError: when no osci.yaml was found in the expected
+                              location
+    """
+    git_root = git_repo.git.rev_parse('--show-toplevel')
+    fpath = os.path.join(git_root, OSCI_YAML)
+    if os.path.isfile(fpath):
+        return pathlib.Path(fpath)
+    else:
+        raise FileNotFoundError(fpath)
+
+
+def load_osci_yaml(git_repo: git.Repo) -> OsciYamlType:
+    """Load osci.yaml content into memory.
+
+    :param git_repo: git repo of the charm
+    :return: parsed content of osci.yaml
+    """
+    fpath = find_osci_yaml(git_repo)
+    with open(fpath, 'r') as f:
+        return yaml.safe_load(f)
+
+
+def get_charm_name(osci: OsciYamlType) -> str:
+    """Get the charm name defined in osci.yaml.
+
+    :param osci: the osci configuration set in the repo.
+    :returns: the charm name
+    """
+    try:
+        vars_ = get_project_vars(osci)
+        return vars_['charm_build_name']
+    except (TypeError, KeyError) as ex:
+        logger.warning('charm_build_name not found in osci.yaml: %s' % ex)
+        raise CharmNameNotFound()
+    raise CharmNameNotFound()
+
+
+def get_project_vars(osci: OsciYamlType) -> Dict[str, Any]:
+    """Get 'vars' section from osci.yaml.
+
+    :param osci: the osci configuration set in the repo.
+    :returns: vars section
+    """
+    for section in osci:
+        if 'project' not in section:
+            continue
+
+        return section['project']['vars']
```

Comment:
This could, theoretically, return None if the 'project' isn't in the osci iterator.  I tend to format this in the following way:

```python
for section in osci:
    try:
        return section['project']['vars']
    except KeyError:
        pass
else:
    raise Exception('not found ...')
```